tests/kernel/interrupt/src/interrupt_offload.c

```diff
@@ -125,55 +132,70 @@ static void trigger_offload_interrupt(const bool real_irq, void *work)
 
 static void t_running(void *p1, void *p2, void *p3)
 {
-	while (1) {
+	k_sem_give(&sync_sem);
+
+	while (wait_for_end == 0) {
 		orig_t_keep_run = 1;
-		k_usleep(1);
+		k_usleep(150);
 	}
+
+	orig_t_keep_run = 1;
+	k_sem_give(&sync_sem);
 }
 
-static void run_test_offload(int case_type, int real_irq)
+static void init_env(int real_irq)
 {
 	static bool wq_already_start;
-	int thread_prio = 1;
-
-	TC_PRINT("case %d\n", case_type);
 
 	/* semaphore used to sync the end */
-	k_sem_init(&end_sem, 0, NUM_WORK);
+	k_sem_init(&sync_sem, 0, 1);
+	k_sem_init(&end_sem, 0, 1);
 
-	if (offload_job_prio_higher) {
-		/* priority of offload job higher than thread */
-		thread_prio = 2;
-	} else {
-		/* priority of offload job lower than thread */
-		thread_prio = 0;
-	}
+	/* initialize global variables */
+	submit_success = 0;
+	offload_job_cnt = 0;
+	orig_t_keep_run = 0;
+	wait_for_end = 0;
 
+	/* initialize the dynamic interrupt while using it */
 	if (real_irq && !vector_num) {
 		init_dyn_interrupt();
 	}
 
-	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
-			(k_thread_entry_t)t_running,
-			NULL, NULL, NULL,
-			K_PRIO_PREEMPT(thread_prio),
-			K_INHERIT_PERMS, K_NO_WAIT);
+	/* initialize all the k_work */
+	for (int i = 0; i < NUM_WORK; i++) {
+		k_work_init(&offload_work[i], entry_offload_job);
+	}
 
 	/* start a work queue thread if not existing */
 	if (!wq_already_start) {
 		k_work_queue_start(&wq_queue, wq_stack, STACK_SIZE,
-		1, NULL);
+		K_PRIO_PREEMPT(1), NULL);
 
 		wq_already_start = true;
 	}
+}
 
-	/* initialize all the k_work */
-	for (int i = 0; i < NUM_WORK; i++) {
-		k_work_init(&offload_work[i], entry_offload_job);
+static void run_test_offload(int case_type, int real_irq)
+{
+	int thread_prio = K_PRIO_PREEMPT(0);
+
+	/* initialize the global variables */
+	init_env(real_irq);
+
+	/* set priority of offload job higher than thread */
+	if (offload_job_prio_higher) {
+		thread_prio = K_PRIO_PREEMPT(2);
 	}
 
+	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
+			(k_thread_entry_t)t_running,
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(thread_prio),
```

Comment:
We are effectively doing `K_PRIO_PREEMPT(K_PRIO_PREEMPT(priority))` here.

Reply:
Yes, thanks for correcting this. Got it fixed.